presto-main/src/main/java/io/prestosql/sql/planner/optimizations/AddExchanges.java

```diff
@@ -227,7 +227,8 @@ public PlanWithProperties visitAggregation(AggregationNode node, PreferredProper
                         gatheringExchange(idAllocator.getNextId(), REMOTE, child.getNode()),
                         child.getProperties());
             }
-            else if (!child.getProperties().isStreamPartitionedOn(partitioningRequirement) && !child.getProperties().isNodePartitionedOn(partitioningRequirement)) {
+            else if ((!child.getProperties().isStreamPartitionedOn(partitioningRequirement) && !child.getProperties().isNodePartitionedOn(partitioningRequirement)) ||
+                    node.hasEmptyGroupingSet()) {
```

Comment:
should this be applied when `child.getProperties().isNodePartitionedOn(partitioningRequirement) && node.hasEmptyGroupingSet()`?

Reply:
Yes, it has to be applied unconditionally. In the sample query, the subquery is partitioned in a compatible way and it prevents the partial from being pushed down beyond the exchange, which causes the failure.

Reply:
if it is already repartitioned in the compatible way, maybe we don't need a partial?

Reply:
Partials and remote exchanges are required when default aggregation is present. Each partial aggregation produces default value and such rows are then aggregated in final exchange so that only single default value is outputted

Reply:
Makes sense, thanks for explanation. Maybe let's capture this as a code comment.